**Enterprise LLM Usage, Performance & User Satisfaction Dataset**
*Overview*
- This dataset is a synthetic Generative AI (GenAI) and Large Language Model (LLM) usage dataset designed for data science, machine learning, and analytics projects.

- It simulates real-world interactions between users and enterprise LLM applications by capturing information related to model usage, prompt characteristics, response performance, latency, token consumption, hallucinations, costs, and user satisfaction.


*Note:This is a synthetic dataset created for educational, research, and machine learning purposes. It does not contain any real user information or proprietary data.*

[Data source](https://www.kaggle.com/datasets/mirzayasirabdullah07/llm-and-generative-ai-usage-analytics-dataset)


#### **Loading Library**

In [1]:
import pandas as pd

#### **Loading Data**

In [2]:
df = pd.read_csv("data/genai_usage/genai_llm_usage_dataset_1000.csv")

#### **Getting the dimensions of the data**

In [4]:
rows, columns = df.shape 
output = f"Features: {columns} || Records: {rows}"
print(output)

Features: 14 || Records: 1000


note: 
1000 records is littel dat for AI hence we need at least 30k records if that is not possible you can try and use eg bayesian techniques

#### **Get a summary of the structure of the table**

- data types
- Non nulll value counts
- Memory usage

In [5]:
df.dtypes

session_id               int64
model_name                 str
application_domain         str
task_type                  str
prompt_length            int64
total_tokens             int64
temperature            float64
top_p                  float64
rag_enabled              int64
latency_sec            float64
hallucination_flag       int64
user_satisfaction        int64
estimated_cost_usd     float64
successful_response      int64
dtype: object

In [6]:
df.isna().sum()

session_id             0
model_name             0
application_domain     0
task_type              0
prompt_length          0
total_tokens           0
temperature            0
top_p                  0
rag_enabled            0
latency_sec            0
hallucination_flag     0
user_satisfaction      0
estimated_cost_usd     0
successful_response    0
dtype: int64

kindly note, when handling or dealing with missing data you must check why the data is missing then apply the necessary remedies, 

[How to handle missing values](https://www.datacamp.com/tutorial/techniques-to-handle-missing-data-values)

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   session_id           1000 non-null   int64  
 1   model_name           1000 non-null   str    
 2   application_domain   1000 non-null   str    
 3   task_type            1000 non-null   str    
 4   prompt_length        1000 non-null   int64  
 5   total_tokens         1000 non-null   int64  
 6   temperature          1000 non-null   float64
 7   top_p                1000 non-null   float64
 8   rag_enabled          1000 non-null   int64  
 9   latency_sec          1000 non-null   float64
 10  hallucination_flag   1000 non-null   int64  
 11  user_satisfaction    1000 non-null   int64  
 12  estimated_cost_usd   1000 non-null   float64
 13  successful_response  1000 non-null   int64  
dtypes: float64(4), int64(7), str(3)
memory usage: 135.8 KB


# **Data Understanding**

In [8]:
df.describe()

,session_id,prompt_length,total_tokens,temperature,top_p,rag_enabled,latency_sec,hallucination_flag,user_satisfaction,estimated_cost_usd,successful_response
count,1000.000000,1000.000000,1000.000000,1000.00000,1000.00000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,500.500000,1022.780000,1636.763000,0.74728,0.74207,0.502000,5.368310,0.507000,4.345000,0.018026,0.202000
std,288.819436,574.062391,681.390085,0.42545,0.14234,0.500246,2.375312,0.500201,0.581647,0.011971,0.401693
min,1.000000,12.000000,71.000000,0.00000,0.50000,0.000000,0.310000,0.000000,3.000000,0.000500,0.000000
25%,250.750000,520.750000,1118.000000,0.37000,0.62000,0.000000,3.300000,0.000000,4.000000,0.008400,0.000000
50%,500.500000,1056.000000,1635.500000,0.74000,0.74000,1.000000,5.460000,1.000000,4.000000,0.015400,0.000000
75%,750.250000,1493.250000,2142.250000,1.11000,0.86000,1.000000,7.370000,1.000000,5.000000,0.024900,0.000000
max,1000.000000,2000.000000,3158.000000,1.50000,1.00000,1.000000,10.440000,1.000000,5.000000,0.060300,1.000000


**Insights:**

Prompt Length: 
- average promt sizes were at 1022 words
- the prompt sizes had a spread or rather standard deviation of 574
- Lowest prompt size was 12 words
- 25% of the prompt lengths were below 520 words
- 50% of the prompt lengths were below 1056 words
- 75% of the prompt lengths were below 1493 words
- highest promt was 2000 words


In [9]:
df.describe(include='object')

/tmp/ipykernel_15884/87514550.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,model_name,application_domain,task_type
count,1000,1000,1000
unique,6,7,6
top,Mistral Large,Customer Support,RAG
freq,193,165,180


In [11]:
df['application_domain'].value_counts() # gives us a summary of all the unique qualities and their frequencies

application_domain
Customer Support    165
Education           156
Retail              154
Finance             137
Healthcare          131
Coding              130
Legal               127
Name: count, dtype: int64

In [13]:
df['application_domain'].describe()

count                 1000
unique                   7
top       Customer Support
freq                   165
Name: application_domain, dtype: object

In [12]:
df['task_type'].value_counts() # gives us a summary of all the unique qualities and their frequencies

task_type
RAG                180
QA                 176
Translation        175
Code Generation    174
Summarization      156
Classification     139
Name: count, dtype: int64

In [ ]:
Insights: 

- Mistral Large LLM model was the most used wuth a frequency of 193, Qwen 2.5 model came second with 177 usages and Llama 3.1 came in last out of 6 models with a frequency of 148 usages
- Customer Support  was the most applied use_case frequency of 165 queries, Education came second with 156 queries and Legal  came in last out of 7 use cases with a frequency of 127 usages
- RAG  was the most searched task type with a frequency of 180 queries, QA came second with 176 queries and Classification  came in last out of 6 use cases with a frequency of 139 usages

# **Data Preparation**

# **Modeling**

# **Evaluation**

# **Deployment**